weather_dl/download_pipeline/parsers.py

```diff
@@ -69,6 +69,86 @@ def date(candidate: str) -> datetime.date:
     return converted
 
 
+def time(candidate: str) -> datetime.time:
+    """Converts ECMWF-format time strings into a `datetime.time`.
+
+    Accepted time formats:
+    - HH:MM
+    - HHMM
+    - HH
+
+    For example:
+    - 18:00
+    - 1820
+    - 18
+
+    Note: If MM is omitted it defaults to 00.
+    """
+    converted = None
+
+    accepted_formats = ["%H", "%H:%M", "%H%M"]
+
+    for fmt in accepted_formats:
+        try:
+            converted = datetime.datetime.strptime(candidate, fmt).time()
+            break
+        except ValueError:
+            pass
+
+    if converted is None:
+        raise ValueError(
+            f"Not a valid time: '{candidate}'. Please use valid format."
+        )
+
+    return converted
+
+
+def day_month_year(key: str, candidate: t.Any) -> int:
+    """Converts day, month and year strings into 'int'.
+
+    Also validates the day and month input.
+    """
+    converted = None
+
+    def validate(key: str, candidate: int) -> None:
+        nonlocal converted
+        try:
+            if key == "day":
+                assert 1 <= candidate <= 31
+            if key == "month":
+                assert 1 <= candidate <= 12
+        except AssertionError:
+            converted = None
+
+    if (isinstance(candidate, str) or isinstance(candidate, int)):
+        try:
+            converted = int(candidate)
+            validate(key, converted)
+        except ValueError:
+            pass
+
+    if converted is None:
+        raise ValueError(
+            f"Not a valid day or month or year: {candidate}. Please use valid value."
+        )
+
+    return converted
+
+
+def typecast(key: str, value: t.Any) -> t.Any:
+    """typed the value to its appropriate datatype"""
+    if key == "date":
+        return date(value)
+    if key == "time":
+        return time(value)
+    if key == "day" or key == "month" or key == "year":
+        return day_month_year(key, value)
```

Comment:
I see why you recommended using dictionary-based control flow in standup the other day. Now that I see it all line up, I agree that this would make a bit cleaner.